clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -1491,20 +1495,25 @@ public void setPartition(int partition) {
 
             if (log.isTraceEnabled()) {
                 // Log the message here, because we don't know the partition before that.
-                log.trace("Attempting to append record {} with callback {} to topic {} partition {}", record, userCallback, record.topic(), partition);
+                log.trace("Attempting to append record {} with callback {} to topic {} partition {}", record, userCallback, topic, partition);
             }
+
+            // Reset record to null here so that it doesn't have to be alive as long as the batch is.
+            record = null;
         }
 
         public int getPartition() {
             return partition;
         }
 
         public TopicPartition topicPartition() {
-            if (record == null)
-                return null;
-            return partition == RecordMetadata.UNKNOWN_PARTITION
-                    ? ProducerInterceptors.extractTopicPartition(record)
-                    : new TopicPartition(record.topic(), partition);
+            if (partition != RecordMetadata.UNKNOWN_PARTITION)
+                return new TopicPartition(topic, partition);
```

Comment:
It's a bit surprising to allocate every time a method like this is called. Can we not allocate the topic partition once and reuse it?

Reply:
The topicPartition is called once in success case (maybe twice in error case).  I'll add a comment.

Reply:
Even so, the way this method is used can change over time. And then you end up with a lot of unexpected allocation. The way I suggested is more robust.

Reply:
Ok.